sympy/core/numbers.py

```diff
@@ -2422,6 +2422,89 @@ def __floordiv__(self, other):
     def __rfloordiv__(self, other):
         return Integer(Integer(other).p // self.p)
 
+    def __lshift__(self, other):
+        if global_parameters.evaluate:
+            if isinstance(other, int):
+                return Integer(self.p << other)
+            elif isinstance(other, Integer):
+                return Integer(self.p << other.p)
+        return NotImplemented
```

Comment:
I don't think we need to check evaluate here since there is no unevaluated form. This would raise TypeError when evaluate=False which isn't particularly useful and probably contradicts the Integral ABC anyway.

I'd write this as:
```python
if not isinstance(other, (int, Integer)):
    return NotImplemented
return Integer(self.p << int(other))
```
I'm not sure what the numbers ABC say about interaction between different implementations of each ABC. Should this method work for any `numbers.Integral`? If so then maybe it should be:
```python
if not isinstance(other, (int, Integer, Integral)):
    return NotImplemented
return Integer(self.p << int(other))
```

Reply:
Oh! I like that much better. I took a stab with a slight twist on style. (I've developed a habitual nervousness around `if not …` because it tends to mess with my brain, even though it's perfectly readable in this case.) If you prefer the flow as you've written it, let me know and I'll change it. I also like including `Integral`, so I've done that as well.

Technically, one could probably get away with `isinstance(other, (Integer, numbers.Integral))`, since `isinstance(int(0), numbers.Integral)` should evaluate to `True`.

Reply:
> Technically, one could probably get away with `isinstance(other, (Integer, numbers.Integral))`, since `isinstance(int(0), numbers.Integral)` should evaluate to `True`.

Wouldn't the changes here also mean that `isinstance(Integer(0), numbers.Integral` would be True as well?

I doubt performance is an issue here (anyone looking for better performance should just use `int` rather than `Integer`) but testing `isinstance` with an ABC is slow so it's faster if you can shortcut the common cases:
```python
In [119]: from numbers import Integral

In [120]: x = 1

In [121]: %timeit isinstance(x, int)
218 ns ± 0.917 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [122]: %timeit isinstance(x, Integral)
1.64 µs ± 74 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [123]: %timeit isinstance(x, (int, Integral))
269 ns ± 36.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
Interesting. I could make one slight tweak to remove `Integer` from the `isinstance` checks in the `__r…__` methods (to signal that `other` won't be an `Integer` on that path), but I think its presence is logically harmless.

This is an example of what I mean:

```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index af20ef64e3..73f828de4e 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -2429,7 +2429,7 @@ def __lshift__(self, other):
             return NotImplemented

     def __rlshift__(self, other):
-        if isinstance(other, (int, Integer, numbers.Integral)):
+        if isinstance(other, (int, numbers.Integral)):
             return Integer(int(other) << self.p)
         else:
             return NotImplemented
```

Just let me know what you'd prefer.

Reply:
My preference would be to test for Integer explicitly so that it is very clear that this works with Integer but honestly I don't think it matters much so I don't mind.

A more important confusion that someone would likely have when looking at this code is to wonder why `sympify` isn't being used. There should be a long comment above these methods saying something like:

Bitwise operations are defined for Integer only and not for general sympy expressions. This is for compatibility with the numbers.Integer ABC which only defines these operations between instances of numbers.Integral. These methods should check explicitly for integer types rather than using sympify because they should not accept arbitrary symbolic expressions and there is no symbolic analogue of the bitwise operations.

Reply:
Added. I tweaked the language _slightly_, but reproduced it largely verbatim.

Reply:
> I don't think we need to check evaluate here since there is no unevaluated form.
```python
>>> Rational(4, 8, gcd=1)
4/8
```

Reply:
@smichr, does this mean I should re-introduce the `if global_parameters.evaluate: …` check from 82f7de3?